aws-java-sdk-core/src/main/java/com/amazonaws/auth/profile/internal/ProfilesConfigFileLoader.java

```diff
@@ -82,43 +85,103 @@
+                    String.format("Unable to load credentials into profile [%s]: AWS Secret Access Key is not specified.",
+                            profileName));
+        }
+
+        if (sessionToken == null) {
+            return new Profile(profileName,
+                    new BasicAWSCredentials(accessKey, secretKey));
+        } else {
+            if (sessionToken.isEmpty()) {
                 throw new AmazonClientException(
-                        String.format("Unable to load credentials into profile [%s]: AWS Secret Access Key is not specified.",
+                        String.format("Unable to load credentials into profile [%s]: AWS Session Token is empty.",
                                 profileName));
             }
 
-            if (sessionToken == null) {
-                profilesByName.put(
-                        profileName,
-                        new Profile(profileName,
-                                    new BasicAWSCredentials(accessKey, secretKey)));
-            } else {
-                if (sessionToken.isEmpty()) {
-                    throw new AmazonClientException(
-                            String.format("Unable to load credentials into profile [%s]: AWS Session Token is empty.",
-                                    profileName));
-                }
-
-                profilesByName.put(
-                        profileName,
-                        new Profile(profileName,
-                                    new BasicSessionCredentials(accessKey, secretKey, sessionToken)));
-            }
+            return new Profile(profileName,
+                    new BasicSessionCredentials(accessKey, secretKey, sessionToken));
         }
+    }
 
-        return profilesByName;
+    private static Profile fromAssumeRole(String profileName, Map<String, String> properties, Map<String, Map<String, String>> allProfileProperties, ProfileCredentialsService profileCredentialsService) {
+        String roleArn = properties.get(Profile.ROLE_ARN);
+        String sourceProfileName = properties.get(Profile.SOURCE_PROFILE);
+        String roleSessionName = properties.get(Profile.ROLE_SESSION_NAME);
+        String externalId = properties.get(Profile.EXTERNAL_ID);
+
+        if (sourceProfileName == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load credentials into profile [%s]: Source profile name is not specified",
+                            profileName));
+        }
+
+        Map<String, String> sourceProfileProperties = allProfileProperties.get(sourceProfileName);
+
+        if (sourceProfileProperties == null) {
+            throw new AmazonClientException(
+                    String.format("Unable to load source profile [%s]: Source profile not found [%s]",
+                            profileName, sourceProfileName));
+        }
+
+        String sourceAccessKey = sourceProfileProperties.get(Profile.AWS_ACCESS_KEY_ID);
```

Comment:
There are a few issues:

1. The source profile may not have been parsed yet, so may not be in the profilesByName map.
2. The source profile can not be another role based profile. That is partly because it is the way botocore implements it and partly because then you could have a chain of role assumptions. It would require some cycle checks and generally be more complex for little gain.